### ISSUE TITLE
*.pp contentFiles not copied to output directory in 3.0 SDK

### ISSUE THREAD

**Issue by megakid** [dotnet/sdk#3719]
I've been digging into this issue whereby a `.pp` file inside a package `contentFiles` (with `copyToOutput="true"`) via `PackageReference` is not copied into the output directory during a `dotnet publish` / `dotnet build`.  This works fine with .NET Core SDK `2.2.*` but is broken in `3.0.100`.  Non-`.pp` files work fine.

I have debugged the MSBuild process and can see that the files are transformed successfully but they do not get copied to the output unlike non-`.pp` files.

My test project/package structure is basic and looks like this:

`[Library A (TestAbc)] <== (PackageReference) == [Exe B]`

Library A's `.nuspec`:

```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>TestAbc</id>
    <version>160.0.0</version>
    <authors>TestAbc</authors>
    <owners>TestAbc</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <description>Package Description</description>
    <dependencies>
      <group targetFramework=".NETFramework4.7.1" />
    </dependencies>
    <contentFiles>
      <files include="any/any/Test.ps1.pp" buildAction="Content" copyToOutput="true" />
      <files include="any/any/NonTransformed.ps1" buildAction="Content" copyToOutput="true" />
    </contentFiles>
  </metadata>
</package>
```

Exe B's `csproj` (I've also tried targeting `netcoreapp3.0`)

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>net472</TargetFrameworks>
    <OutputType>Exe</OutputType>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="TestAbc" Version="163.0.0" />
  </ItemGroup>
</Project>
```

My dotnet version:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional>dotnet --version
3.0.100
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional>dotnet --list-sdks
3.0.100 [C:\Program Files\dotnet\sdk]
```

My build command - I couldn't figure out how to get binlogs from msbuild via the dotnet publish command so I used the `CopyLocalLockFileAssemblies` property workaround although I'm not sure this is required with 3.0 SDK (see [here](https://github.com/dotnet/sdk/pull/2646)):

`C:/Program Files/dotnet/dotnet.exe --diagnostics msbuild /target:Build /property:PackageOutputPath=C:/Projects/test/bin/packages /property:IncludeSymbols=true /property:IncludeSource=true /property:CopyLocalLockFileAssemblies=true /property:Version=1.0.0 /property:configuration=Release "C:/Projects/squid2/src/ExeA/ExeA.csproj" -restore /bl:build.binlog -v:diag`

After this command I see `NonTransformed.ps1` in the output folder but not `Test.ps1` (or `Test.ps1.pp`).
I would expect to see both `.ps1`s in my output folder at that point.

Here's my debugging so far:

I see the .pp file getting picked up:

![image](https://user-images.githubusercontent.com/1478003/66257599-1d579a80-e793-11e9-9f82-74730dc52175.png)

It gets copied to a temp directory without the .pp extension (`obj\Release\net472\NuGet\9374D6CD1CF741AA241FA0906DC657D57C8CB116\TestAbc\165.0.0\Test.ps1` in this case).

Then I see the preprocessor task running and see that the file has now been overwritten/updated with the template tokens replaced.
 
![image](https://user-images.githubusercontent.com/1478003/66257582-f600cd80-e792-11e9-8a0b-9a481c9bab8a.png)

I notice later in the process that the 2 files are listed side by side but the `TargetPath` looks wrong (to my untrained eye) under the `Test.ps1` element - first difference is the content path is an absolute path for the `NonTransformed.ps1` but a relative path for `Test.ps1`.  The task `AssignTargetPath` is used to generate a `TargetPath` (see [here](https://github.com/microsoft/msbuild/blob/vs16.3/src/Tasks/AssignTargetPath.cs)) which generates the correct one on `NonTransformed.ps1` but not `Test.ps1`.  **UPDATE** From tracing the 2.2 build, I can see it is the same in 2.2 so this is fine.

![image](https://user-images.githubusercontent.com/1478003/66257667-fd74a680-e793-11e9-9b07-28edda0a4a23.png)

I can't figure out where `Test.ps1` is being dropped but later on I clearly see `NonTransformed.ps1` alone in the copy task:

![image](https://user-images.githubusercontent.com/1478003/66257561-c94cb600-e792-11e9-8f5d-f8bd7aed22e6.png)

My output dir afterwards:

![image](https://user-images.githubusercontent.com/1478003/66258015-0e271b80-e798-11e9-919b-c90d8b39c2bf.png)

Exact same build but with `SDK 2.2.402`:

![image](https://user-images.githubusercontent.com/1478003/66258002-e46df480-e797-11e9-9d34-cdae57598bac.png)

I've included my 2.2 and 3.0 binlogs here for comparison:

[msbuild-binlogs.zip](https://github.com/dotnet/sdk/files/3693524/msbuild-binlogs.zip)

**Comment by lscpike** [comment]
Digging through the code a little it seems the _ContentCopyLocalItems is no longer used. That seems to be where the .pp file output was previously added to the ReferenceCopyLocalPaths.

See https://github.com/dotnet/sdk/commit/1936ebd7307cebd6c5d1d8d699b9844f73b8fb2c#diff-726c4af93875efa855c29e94f7c0b33eL498

Feels like an accidental bug to me.

**Comment by livarcocc** [comment]
@sfoslund can you work with @dsplaisted to come up with a fix proposal for this?

**Comment by remcoros** [comment]
@megakid did you come up with a good work-around already? 

We are using .pp files for non-source files (.txt.pp files) and relied upon the "copy to output" functionality. These files are now missing in our builds when using the 3.0 sdk.

**Comment by megakid** [comment]
No work around that I could find.

We ended up moving away from .pp files as a workaround but obviously that
might not be an option for many. IMO this is a major breaking change for
those using .pp files - the fact it fails so silently means a very
frustrating experience.

**Comment by remcoros** [comment]
I managed to get it working with Sdk 3.0, haven't tested this with Sdk < 3.0

Adding this in a [package].targets file, including it in the nuget package, so it auto-runs on build in the consuming project:

```xml
  <Target Name="_FixSdk30ReferenceCopyLocalPaths" AfterTargets="RunProduceContentAssets">
    <ItemGroup>
      <ReferenceCopyLocalPaths Include="@(_ContentCopyLocalItems)" />
    </ItemGroup>
  </Target>
```

**Comment by sfoslund** [comment]
Addressed in #3771
